From 3da934e6ca914a35130eafcbd19d252e8d666914 Mon Sep 17 00:00:00 2001
From: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
Date: Wed, 5 Jun 2024 06:32:13 +0530
Subject: [PATCH] hostapd: Add power mode to channel switch command

Add 6 Ghz power mode to channel switch command, to change the
channel and power mode in a single command atomically.

Patch dependency: n00-001-hostapd-add-support-for-6GHz-operation.patch
Tag: 6G_REGULATORY

Signed-off-by: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
---
 hostapd/ctrl_iface.c         |  8 ++++++++
 src/ap/ctrl_iface_ap.c       | 15 +++++++++++++++
 src/ap/drv_callbacks.c       | 23 +++++++++++++++++++++++
 src/ap/hostapd.c             |  6 +++++-
 src/ap/interference.c        |  1 +
 src/drivers/driver.h         |  2 ++
 src/drivers/driver_nl80211.c | 11 +++++++++--
 7 files changed, 63 insertions(+), 3 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 85142bc..acdbbb3 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -2990,6 +2990,14 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 		return ret;
 	}
 
+	if (iface->power_mode_6ghz_before_change > -1) {
+		wpa_printf(MSG_ERROR, "Power mode change in progress");
+		return -1;
+	}
+
+	if (settings.power_mode > -1)
+		iface->power_mode_6ghz_before_change = settings.power_mode;
+
 	ret = hostapd_check_validity_device_params(&settings.freq_params);
 	if (ret) {
 		wpa_printf(MSG_ERROR, "chanswitch: invalid device parameters provided %d %d",
diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 72821b4..31b53a1 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1363,6 +1363,7 @@ int hostapd_parse_csa_settings(const char *pos,
 
 	os_memset(settings, 0, sizeof(*settings));
 	settings->cs_count = strtol(pos, &end, 10);
+	settings->power_mode = -1;
 	if (pos == end) {
 		wpa_printf(MSG_ERROR, "chanswitch: invalid cs_count provided");
 		return -1;
@@ -1400,6 +1401,20 @@ int hostapd_parse_csa_settings(const char *pos,
 	SET_CSA_SETTING_EXT(punct_bitmap);
 	SET_CSA_SETTING(bandwidth_device);
 	SET_CSA_SETTING(center_freq_device);
+	SET_CSA_SETTING_EXT(power_mode);
+	if (!is_6ghz_freq(settings->freq_params.freq) &&
+	    (settings->power_mode != -1)) {
+		wpa_printf(MSG_ERROR,
+			   "chanswitch: power mode is not supported for non- 6 GHz frequency");
+		return -1;
+	}
+
+	if (settings->power_mode < -1 ||
+	    settings->power_mode > HE_6GHZ_VERY_LOW_POWER_AP) {
+		wpa_printf(MSG_ERROR, "chanswitch: invalid 6 GHz power_mode provided");
+		return -1;
+	}
+
 	settings->freq_params.ht_enabled = !!os_strstr(pos, " ht");
 	settings->freq_params.vht_enabled = !!os_strstr(pos, " vht");
 	settings->freq_params.he_enabled = !!os_strstr(pos, " he");
diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 2318acc..c82ec67 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -924,6 +924,21 @@ void hostapd_event_sta_opmode_changed(struct hostapd_data *hapd, const u8 *addr,
 	}
 }
 
+static int hostapd_switch_power_mode(struct hostapd_data *hapd)
+{
+	struct he_6ghz_pwr_mode_settings settings;
+
+	settings.pwr_mode =
+		hapd->iface->power_mode_6ghz_before_change;
+	hapd->iface->power_mode_6ghz_before_change = -1;
+	if (hostapd_drv_set_6ghz_pwr_mode(hapd, &settings)) {
+		wpa_msg(hapd->msg_ctx, MSG_ERROR, AP_CSA_FINISHED
+			"Power mode change failed");
+		return -EINVAL;
+	}
+
+	return 0;
+}
 
 void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 			     int offset, int width, int cf1, int cf2,
@@ -1100,6 +1115,14 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 
 	if (hapd->csa_in_progress &&
 	    freq == hapd->cs_freq_params.freq) {
+		if (hapd->iface->power_mode_6ghz_before_change > -1) {
+			if (hapd->iface->power_mode_6ghz_before_change !=
+			    hapd->iconf->he_6ghz_reg_pwr_type) {
+				if (hostapd_switch_power_mode(hapd))
+					wpa_printf(MSG_ERROR, "Power mode change failed");
+			}
+		}
+
 		hostapd_cleanup_cs_params(hapd);
 		hapd->disable_cu = 1;
 		ieee802_11_set_beacon(hapd);
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index e5115de..1f70c6b 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4531,7 +4531,7 @@ static int hostapd_fill_csa_settings(struct hostapd_data *hapd,
 #ifdef CONFIG_IEEE80211BE
 	u16 old_punct_bitmap;
 #endif /* CONFIG_IEEE80211BE */
-	u8 chan, bandwidth;
+	u8 chan, bandwidth, old_reg_6ghz_power_mode;
 	u8 oper_centr_freq0_idx = 0;
 	u8 oper_centr_freq1_idx = 0;
 	int sec_channel_offset = settings->freq_params.sec_channel_offset;
@@ -4605,6 +4605,8 @@ static int hostapd_fill_csa_settings(struct hostapd_data *hapd,
 	if (ret)
 		return ret;
 
+	old_reg_6ghz_power_mode = iface->conf->he_6ghz_reg_pwr_type;
+	iface->conf->he_6ghz_reg_pwr_type = settings->power_mode;
 #ifdef CONFIG_IEEE80211BE
 	old_punct_bitmap = iface->conf->punct_bitmap;
 	iface->conf->punct_bitmap = settings->punct_bitmap;
@@ -4665,6 +4667,7 @@ static int hostapd_fill_csa_settings(struct hostapd_data *hapd,
 	 hapd->iface->conf->he_mu_edca.he_qos_info &= 0xfff0;
 	 hapd->parameter_set_count = 0;
 
+	iface->conf->he_6ghz_reg_pwr_type = old_reg_6ghz_power_mode;
 	ret = hostapd_build_beacon_data(hapd, &settings->beacon_csa);
 	if (ret) {
 		free_beacon_data(&settings->beacon_after);
@@ -4702,6 +4705,7 @@ void hostapd_cleanup_cs_params(struct hostapd_data *hapd)
 	hapd->csa_in_progress = 0;
 	hapd->cs_c_off_ecsa_beacon = 0;
 	hapd->cs_c_off_ecsa_proberesp = 0;
+	hapd->iface->power_mode_6ghz_before_change = -1;
 }
 
 
diff --git a/src/ap/interference.c b/src/ap/interference.c
index d6f0d7d..f1fe614 100644
--- a/src/ap/interference.c
+++ b/src/ap/interference.c
@@ -499,6 +499,7 @@ int hostapd_intf_awgn_detected(struct hostapd_iface *iface, int freq, int chan_w
 	settings.freq_params.vht_enabled = iface->conf->ieee80211ac;
 	settings.freq_params.he_enabled = iface->conf->ieee80211ax;
 	settings.freq_params.eht_enabled= iface->conf->ieee80211be;
+	settings.power_mode = -1;
 
 	for (i = 0; i < iface->num_bss; i++) {
 		/* Save CHAN_SWITCH VHT and HE config */
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index f888319..fa7bbb9 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2819,6 +2819,7 @@ struct beacon_data {
  * @counter_offset_beacon: Offset to the count field in beacon's tail
  * @counter_offset_presp: Offset to the count field in probe resp.
  * @punct_bitmap - Preamble puncturing bitmap
+ * @power_mode: 6 GHz Power mode
  */
 struct csa_settings {
 	u8 cs_count;
@@ -2837,6 +2838,7 @@ struct csa_settings {
 
 	/* critical_update_flag - critical update flag*/
 	bool beacon_after_cu;
+	int power_mode;
 };
 
 /**
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 0b26b47..24144a9 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -11325,7 +11325,7 @@ static int nl80211_switch_channel(void *priv, struct csa_settings *settings)
 	u8 critical_update = 0;
 
 	wpa_printf(MSG_DEBUG,
-		   "nl80211: Channel switch request (cs_count=%u block_tx=%u freq=%d channel=%d sec_channel_offset=%d width=%d cf1=%d cf2=%d puncturing_bitmap=0x%04x%s%s%s  puncturing bitmap=0x%04x, ru_punct_ofdma=%u)",
+		   "nl80211: Channel switch request (cs_count=%u block_tx=%u freq=%d channel=%d sec_channel_offset=%d width=%d cf1=%d cf2=%d puncturing_bitmap=0x%04x%s%s%s  puncturing bitmap=0x%04x, ru_punct_ofdma=%u power_mode=%u)",
 		   settings->cs_count, settings->block_tx,
 		   settings->freq_params.freq,
 		   settings->freq_params.channel,
@@ -11338,7 +11338,8 @@ static int nl80211_switch_channel(void *priv, struct csa_settings *settings)
 		   settings->freq_params.vht_enabled ? " vht" : "",
 		   settings->freq_params.he_enabled ? " he" : "",
 		   settings->freq_params.ru_punct_bitmap,
-		   settings->freq_params.ru_punct_ofdma);
+		   settings->freq_params.ru_punct_ofdma,
+		   settings->power_mode);
 
 	if (!(drv->capa.flags & WPA_DRIVER_FLAGS_AP_CSA)) {
 		wpa_printf(MSG_DEBUG, "nl80211: Driver does not support channel switch command");
@@ -11412,6 +11413,12 @@ static int nl80211_switch_channel(void *priv, struct csa_settings *settings)
 			 settings->punct_bitmap)))
 		goto error;
 
+	if (is_6ghz_freq(settings->freq_params.freq) && settings->power_mode > -1) {
+		if (nla_put_u8(msg, NL80211_ATTR_6G_REG_POWER_MODE,
+			       settings->power_mode))
+			goto error;
+	}
+
 	/* beacon_after params */
 	ret = set_beacon_data(msg, &settings->beacon_after, false);
 	if (ret)
-- 
2.34.1

